Feature: Infinite money mode

merni-ns opened the pull request:
## Motivation / Problem

Many players want to play in a "sandbox" style where they don't care about the money-making mechanics and just want to build cool networks. At present, this has to be achieved by cheating arbitrary amounts of money, which is an unsatisfactory and clunky solution, apart from messing up the company financial statements.


## Description
This PR adds a setting called "Infinite money mode". When enabled, it allows any command to be performed even if the company's money balance is less than the amount required (or even negative). It also disables the company bankruptcy mechanism, and replaces the balance in the status bar with "infinite money". The actual balance is still shown in the finance window. 

This approach has the advantage that it maintains the integrity of the company financial statements. It charges the company all costs, and does not add any extra arbitrary income (like the money cheat does). It also removes the need for occasionally cheating more money or otherwise infusing money -- players can just forget about the existence of the money mechanism for the duration of the game. Furthermore, it works perfectly fine with AI and multiplayer.

Screenshot:
![image](https://github.com/OpenTTD/OpenTTD/assets/66267867/41d64e24-7887-4416-965d-1593fbd109f8)

## Limitations
- It takes effect on all companies, even AI. At present no check is done whether the company is a human or AI one.
- Looking at vehicle, company, station, etc income statistics, still shows the "real" negative values which (says @TrueBrain) might put off players. (But it is anyway the same at present except for companies)
- Not tested in multiplayer.
- There may be some other edge case where some command/function directly checks how much money a company has instead of going through functions such as `GetAvailableMoneyForCommand`, `CheckCompanyHasMoney`, etc. I have done my best to find and fix all such instances, but there is no guarantee that one more is not lurking somewhere.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

merni-ns committed:
Feature: Infinite money mode

merni-ns committed:
Change: Don't check for sufficient money if infinite money setting is on

merni-ns committed:
Change: Don't check for company bankruptcy if infinite money setting is on

James103 commented on src/saveload/saveload.h:
```diff
@@ -371,6 +371,7 @@ enum SaveLoadVersion : uint16_t {
 	SLV_ECONOMY_DATE,                       ///< 326  PR#10700 Split calendar and economy timers and dates.
 	SLV_ECONOMY_MODE_TIMEKEEPING_UNITS,     ///< 327  PR#11341 Mode to display economy measurements in wallclock units.
 	SLV_CALENDAR_SUB_DATE_FRACT,            ///< 328  PR#11428 Add sub_date_fract to measure calendar days.
+	SLV_IGNORE_BANKRUPTCY,                  ///< 329  PR#..... Add setting to ignore bankruptcy
```
```suggestion
	SLV_IGNORE_BANKRUPTCY,                  ///< 329  PR#11902 Add setting to ignore bankruptcy
```

merni-ns replied:
Ah yes. :facepalm:

merni-ns committed:
Change: Don't display current balance if infinite money is on

frosch123 commented on src/saveload/saveload.h:
```diff
@@ -371,6 +371,7 @@ enum SaveLoadVersion : uint16_t {
 	SLV_ECONOMY_DATE,                       ///< 326  PR#10700 Split calendar and economy timers and dates.
 	SLV_ECONOMY_MODE_TIMEKEEPING_UNITS,     ///< 327  PR#11341 Mode to display economy measurements in wallclock units.
 	SLV_CALENDAR_SUB_DATE_FRACT,            ///< 328  PR#11428 Add sub_date_fract to measure calendar days.
+	SLV_INFINITE_MONEY,                     ///< 329  PR#11902 Add infinite money mode
```
I don't think adding a new setting requires a savegame bump. There are no conversions.

TrueBrain replied:
And adding of settings no longer needs a savegame bump either. It used to, but we solved that a few version ago.

merni-ns replied:
Removed the savegame bump that I added thanks to [outdated docs on the wiki](https://wiki.openttd.org/en/Archive/Source/OpenTTDDevBlackBook/Settings/AddSetting).

2TallTyler commented on src/company_cmd.cpp:
```diff
@@ -196,11 +196,12 @@ void InvalidateCompanyWindows(const Company *company)
 bool CheckCompanyHasMoney(CommandCost &cost)
 {
-	if (cost.GetCost() > 0) {
+	if (cost.GetCost() > 0 && !(_settings_game.difficulty.infinite_money)) {
```
This would be nicer as an early return, as you did in `CompanyCheckBankrupt()`.

merni-ns replied:
I did this and moved the cost condition also to an early return.

2TallTyler reviewed (commented):
The commit prefix for this one should definitely be `Feature:`. I suspect there will be a fair amount of excitement for this one. 😄 Other than that and my one code nitpick, it looks good!

2TallTyler reviewed (commented):
Sorry, just noticed... All of our .ini files end with an empty line. Add that back and I'll approve this! 😄

merni-ns committed:
Codechange: Move GetAvailableMoneyForCommand to company_cmd.cpp and add GetAvailableMoney

The function belongs better in company_cmd.cpp along with the similar CheckCompanyHasMoney and SubtractMoneyFromCompany.
The generalised function GetAvailableMoney is added which allows finding the money of any company, not just the current one. This will be useful in the next commit.

2TallTyler commented on src/command.cpp:
```diff
@@ -97,7 +97,7 @@ inline constexpr auto MakeCommandsFromTraits(std::integer_sequence<T, i...>) noe
 static constexpr auto _command_proc_table = MakeCommandsFromTraits(std::make_integer_sequence<std::underlying_type_t<Commands>, CMD_END>{});
 
 
-/*!
```
Changing comments of unrelated functions should be in a separate PR. 🙂
